### PR TITLE
Better parameter-based tests for AMM liquidity

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
## Description

This is an update to #4533, to better demonstrate the intent of the underlying #4519.

I've pushed changes to #4519, but haven't rebased #4533.

### CODE REVIEW

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+      ({ poolX, poolY, giveX, giveY }) => {
+        const { targetX, targetY } = balancesToReachRatio(
+          poolX,
+          poolY,
+          giveX,
+          giveY,
+        );
+
+        if (targetX.value === 0n && targetY.value === 0n) {
+          t.pass();
```
Comment (dckc) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: It's hard to tell whether this is correct because the spec hasn't been updated.

In particular, an implementation that just always returns `{ targetX: 0n, targetY: 0n }` would pass this test, I think.

I think it's better to filter the _inputs_ to `balancesToReachRatio`. That is: test it on arbitrary _balanced_ requests.

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        } else {
+          const targetProduct = Number(multiply(targetX.value, targetY.value));
+          const withinRange = withinEpsilon(
+            targetProduct,
+            Number(multiply(poolX.value, poolY.value)),
+            0.05,
+          );
```
Comment (dckc) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: One of my goals in doing property-based testing (that's the term of art, btw; not parameter-based) was to avoid thinking about floating point arithmetic. Again, the spec (JSDoc) hasn't been updated to say anything about conversion to floating point, so it's hard for me to evaluate whether this test is correct.

**Review by dckc — CHANGES_REQUESTED** [review]
It looks like there's implicitly a new spec, but to know whether we're testing to it, I need to see an explicit version of it; i.e. an update to the JSDoc for the function under test.

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: I added some details to the description of `balancesToReachRatio`, and filtered inputs.

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: The test now only converts to `Number` in order to create a ratio in the parameter for the call to `imbalancedRequest()` so the ratio doesn't coerce to zero if it's less than 1. `imbalancedRequest()`  uses  `Number(startK)` when multiplying by desiredRatio (a fraction that is less than 1.)

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  balancesToReachRatio,
+} from '../../../src/vpool-xyk-amm/addLiquidity.js';
+
+const { add, multiply } = natSafeMath;
+
+const { brand: brandX } = makeIssuerKit('X');
+const { brand: brandY } = makeIssuerKit('Y');
+
+const oneB = 1_000_000_000n;
+// const dec18 = 1_000_000_000_000_000_000n; // 18 decimals as used in ETH
```
Comment (dckc) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: Try taking out `oneB` and all the `max: oneB` stuff, as well as this dec18 comment?

It was prompted by failing tests. The original thought was that `oneB * dec18` would represent a practical limit: 1B ETH, with 10^18 wei to the ETH. Then that still failed, so I went down to just 1B. Given 10^6 urun to the RUN, that only covers up to 10^3 RUN, which is clearly not enough to cover all intended use of the AMM.

Now that the implementation covers the edge cases, we shouldn't need this limit.

Not critical, I guess.

**Review by dckc — APPROVED** [review]
looks good.

I think we can lose the `max: oneB` stuff, though.

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: We had tried an earlier round of property-based testing, and ended up removing it because it was too expensive to include in tests. Is your expectation that this would be checked in?

Comment (dckc) [reply packages/run-protocol/test/amm/vpool-xyk-amm/test-amm-calc.js]: I have a mild preference for having this checked in.

There is now precedent in ERTP for prroperty-based testing. For straightforward calculation functions like this, I don't notice much cost. But that doesn't mean it's not there...

**Comment by Chris-Hibbert** [comment]
Merged as part of #4519
